keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -382,10 +388,107 @@ class ClusterHandler(
     return spec
   }
 
-  private fun List<ResourceDiff<ServerGroup>>.createsNewServerGroups() =
-    any {
-      !it.isCapacityOrAutoScalingOnly()
+  override suspend fun actuationInProgress(resource: Resource<ClusterSpec>) =
+    resource
+      .spec
+      .locations
+      .regions
+      .map { it.name }
+      .any { region ->
+        orcaService
+          .getCorrelatedExecutions("${resource.id}:$region")
+          .isNotEmpty()
+      }
+
+  /**
+   * Retrieve server group entity tags from clouddriver to identity the "source" of the server group (pipeline or
+   * orchestration), then retrieve execution details from orca to determine the deployment strategy used for that
+   * server group.
+   */
+  private suspend fun ServerGroup.discoverDeploymentStrategy(): ClusterDeployStrategy? {
+    return coroutineScope {
+      val entityTags = async {
+        log.debug("Looking for entity tags on server group $name in application ${moniker.app}, " +
+          "account ${location.account} in search of pipeline/task correlation.")
+        cloudDriverService.getEntityTags(
+          cloudProvider = "aws",
+          account = location.account,
+          application = moniker.app,
+          entityType = "servergroup",
+          entityId = name
+        )
+      }.await()
+
+      if (entityTags.isEmpty()) {
+        log.warn("Unable to find entity tags for server group $name in application ${moniker.app}, " +
+          "account ${location.account}.")
+        return@coroutineScope null
+      }
+
+      val spinnakerMetadata = entityTags.first().tags
+        .find { it.name == "spinnaker:metadata" }
+        ?.let {
+          if (it.value !is Map<*, *>) {
+            null
+          } else {
+            it.value as Map<*, *>
+          }
+        }
+
+      if (spinnakerMetadata == null ||
+        spinnakerMetadata["executionType"] == null ||
+        spinnakerMetadata["executionId"] == null
+      ) {
+        log.warn("Unable to find Spinnaker metadata for server group $name in application ${moniker.app}, " +
+          "account ${location.account} in entity tags.")
+        return@coroutineScope null
+      }
+
+      val executionType = spinnakerMetadata["executionType"]!!.toString()
+      val executionId = spinnakerMetadata["executionId"]!!.toString()
+      val execution = async {
+        when (executionType) {
+          "orchestration" -> orcaService.getOrchestrationExecution(executionId)
+          "pipeline" -> orcaService.getPipelineExecution(executionId)
+          else -> throw SystemException("Unsupported execution type $executionType in Spinnaker metadata.")
+        }
+      }.await()
+
+      // get context from the appropriate execution stage for the execution type, drilling down into the data as needed
+      val context = if (executionType == "pipeline") {
```

Comment:
Given the differences in logic for pipelines vs tasks, I would find it easier to read if there were separate `pipeline` and `task` helper functions, rather than having multiple branches in the method.

Reply:
Makes sense. Will refactor.